### PR TITLE
[Pipelines] Perform mergefunc after constmerge

### LINKED ISSUE #92201: MergeFunction can't merge functions because of jump tables
i observed MergeFunc missing merging opportunities due to switch jump tables.

workarounds i can think of:
* -fno-jump-tables
* move mergefunc before switch jump tables transformation
* move mergefunc after constmerge, which would merge switch jump tables

### PR DESCRIPTION
**Pull request by yamt**
Constmerge can fold switch jump tables, possibly making functions identical again. It can help mergefunc.
On the otherhand, the opposite seems unlikely.

Fixes https://github.com/llvm/llvm-project/issues/92201

### CODE REVIEW

**Comment by yamt** [comment]
@nikic can you review?

(i'm asking you because you commented on https://github.com/llvm/llvm-project/issues/92201 and it seems that you have reviewed similar changes in the past.)

**Review by nikic — COMMENTED** [review]
Can you please add a test in llvm/test/Transforms/PhaseOrdering (e.g. with `-O2 -enable-merge-functions`)?

**Comment by yamt** [comment]
> Can you please add a test in llvm/test/Transforms/PhaseOrdering (e.g. with `-O2 -enable-merge-functions`)?

i added a few tests.

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll [hunk 1]
```diff
@@ -0,0 +1,67 @@
+; RUN: opt -passes="default<O3>" -enable-merge-functions -S < %s | FileCheck %s
```
Comment (nikic) [inline llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll:1]: Use update_test_checks.py please.

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-apple-macosx12.0.0"
+
+; Function Attrs: noinline nounwind optsize ssp uwtable
+define i32 @f(i32 noundef %x) #0 {
+; CHECK-LABEL: @f(
+entry:
+  %x.addr = alloca i32, align 4
+  store i32 %x, ptr %x.addr, align 4, !tbaa !5
```
Comment (nikic) [inline llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll:11]: Drop tbaa metadata, it should not be necessary.

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions3.ll [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  %switch.gep = getelementptr inbounds [8 x i32], ptr @switch.table.g, i64 0, i64 %1
+  %switch.load = load i32, ptr %switch.gep, align 4
+  br label %sw.epilog
+
+sw.epilog:                                        ; preds = %entry, %switch.lookup
+  %x.addr.0 = phi i32 [ %switch.load, %switch.lookup ], [ 0, %entry ]
+  ret i32 %x.addr.0
+}
+
+attributes #0 = { mustprogress nofree noinline norecurse nosync nounwind optsize ssp willreturn memory(none) uwtable "frame-pointer"="all" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="penryn" "target-features"="+cmov,+cx16,+cx8,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" "tune-cpu"="generic" }
```
Comment (nikic) [inline llvm/test/Transforms/PhaseOrdering/X86/merge-functions3.ll:47]: Drop attributes.

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll (hunk 1 above)
Comment (yamt) [reply llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll:2]: done

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll (hunk 2 above)
Comment (yamt) [reply llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll]: done

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions3.ll (hunk 3 above)
Comment (yamt) [reply llvm/test/Transforms/PhaseOrdering/X86/merge-functions3.ll]: done

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll [hunk 4]
```diff
@@ -0,0 +1,78 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -passes="default<O3>" -enable-merge-functions -S < %s | FileCheck %s
+
+target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-apple-macosx12.0.0"
+
+; Function Attrs: noinline nounwind optsize ssp uwtable
+define i32 @f(i32 noundef %x) #0 {
```
Comment (nikic) [inline llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll:8]:
```suggestion
define i32 @f(i32 noundef %x) {
```
Here and elsewhere.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll (hunk 4 above)
Comment (yamt) [reply llvm/test/Transforms/PhaseOrdering/X86/merge-functions2.ll]: done. thank you.

**Comment by yamt** [comment]
can this land?

**Pull request merged**